src/sys/sendfile.rs

```diff
@@ -2,12 +2,83 @@ use std::os::unix::io::RawFd;
 use std::ptr;
 
 use libc::{self, off_t};
+#[cfg(target_os = "freebsd")]
+use libc::c_int;
 
 use Result;
+#[cfg(target_os = "freebsd")]
+use sys::uio::IoVec;
 use errno::Errno;
 
+#[cfg(any(target_os = "linux",
+          target_os = "android",
+          all(unix, target_env = "uclibc")))]
 pub fn sendfile(out_fd: RawFd, in_fd: RawFd, offset: Option<&mut off_t>, count: usize) -> Result<usize> {
     let offset = offset.map(|offset| offset as *mut _).unwrap_or(ptr::null_mut());
     let ret = unsafe { libc::sendfile(out_fd, in_fd, offset, count) };
     Errno::result(ret).map(|r| r as usize)
 }
+
+#[cfg(any(target_os = "dragonfly",
+          target_os = "freebsd",
+          target_os = "ios",
+          target_os = "macos"))]
+#[allow(missing_debug_implementations)]
+#[allow(missing_copy_implementations)]
+#[repr(C)]
+pub struct SfHdTr(libc::sf_hdtr);
+
+#[cfg(any(target_os = "dragonfly",
+          target_os = "freebsd",
+          target_os = "ios",
+          target_os = "macos"))]
+impl<'a> SfHdTr {
+    pub fn new(headers: &'a [IoVec<&'a [u8]>], trailers: &'a [IoVec<&'a [u8]>]) -> SfHdTr {
+        SfHdTr(libc::sf_hdtr {
+            headers: { if !headers.is_empty() { headers.as_ptr() as *mut libc::iovec } else { ptr::null_mut() } },
+            hdr_cnt: headers.len() as i32,
+            trailers: { if !trailers.is_empty() { trailers.as_ptr() as *mut libc::iovec} else { ptr::null_mut() } },
+            trl_cnt: trailers.len() as i32
+        })
+    }
+}
+
+#[cfg(target_os = "freebsd")]
+libc_bitflags!{
+    pub struct SfFlags: c_int {
+        SF_NODISKIO;
+        SF_MNOWAIT;
```

Comment:
`SF_MNOWAIT` is obsolete in FreeBSD 11.0 and above.  The last release that makes use of it, 10.4, goes EoL on Oct 31 of this year.  So I'd rather not expose this symbol in Nix unless there is specific demand.